packages/webviz-core/src/dataProviders/MemoryCacheDataProvider.js

```diff
@@ -267,24 +268,33 @@ export default class MemoryCacheDataProvider implements DataProvider {
     this._maybeRunNewConnections();
   }
 
-  _maybeRunNewConnections() {
-    const newConnection = getNewConnection({
-      currentRemainingRange: this._currentConnection ? this._currentConnection.remainingBlockRange : undefined,
-      readRequestRange: this._readRequests[0] ? this._readRequests[0].blockRange : undefined,
-      downloadedRanges: this._getDownloadedBlockRanges(),
-      lastResolvedCallbackEnd: this._lastResolvedCallbackEnd,
-      cacheSize: this._totalNs <= MINIMUM_CACHE_SIZE_NS ? Infinity : Math.ceil(READ_AHEAD_NS / MEM_CACHE_BLOCK_SIZE_NS),
-      fileSize: this._blocks.length,
-      continueDownloadingThreshold: 3, // Somewhat arbitrary number to not create new connections all the time.
-    });
-    if (newConnection) {
-      this._setConnection(newConnection).catch((err) => {
+  async _maybeRunNewConnections() {
+    while (true) {
+      const newConnection = getNewConnection({
+        currentRemainingRange: this._currentConnection ? this._currentConnection.remainingBlockRange : undefined,
+        readRequestRange: this._readRequests[0] ? this._readRequests[0].blockRange : undefined,
+        downloadedRanges: this._getDownloadedBlockRanges(),
+        lastResolvedCallbackEnd: this._lastResolvedCallbackEnd,
+        cacheSize: this._totalNs <= MINIMUM_CACHE_SIZE_NS ? Infinity : READ_AHEAD_BLOCKS,
+        fileSize: this._blocks.length,
+        continueDownloadingThreshold: 3, // Somewhat arbitrary number to not create new connections all the time.
+      });
+      if (!newConnection) {
+        // All read requests done, or there is a connection already in progress handling them.
+        break;
+      }
+      const connectionSuccess = await this._setConnection(newConnection).catch((err) => {
         reportError(
           `MemoryCacheDataProvider connection ${this._currentConnection ? this._currentConnection.id : ""}`,
           err ? err.message : "<unknown error>",
           "app"
         );
       });
+      if (!connectionSuccess) {
+        // Connection interrupted, or otherwise unsuccessful.
+        break;
+      }
+      // See if there are any more read requests we should field.
```

Comment:
Maybe it would be even clearer to call `_resolveFinishedReadRequests` here instead of before the return statements below?

Reply:
(Ah you still need it at the end of the while loop, but maybe the one [here](https://github.com/cruise-automation/webviz/pull/356/commits/d9b3bbbbb44634ee38f844759afd1d9b94dba36f#diff-ff549a830b60d70e1ed51ce42222356fR386) would be clearer when moved? Not sure.)

Reply:
I think that makes sense for the first one, but not for the second. We know that `_setConnection` will return before `this._readRequests[0]` is done, but inside the `while` loop we might also complete *other* read requests.

What do you think about moving it just underneath the call to `_updateProgess` and `_purgeOldBlocks`?

Reply:
Oh yeah that looks like a great place!

Reply:
In fact then you can even put it before `_purgeOldBlocks`. Would that solve the bug you found?

Reply:
Done. I put it above those calls instead of below, because there's a *tiny* chance we might purge something that could be used to resolve a request. (This sadly doesn't fix the other bug.)

Reply:
Haha looks like we're getting some nice race conditions in our comments. 😉